cmd/console/deploy/deploy.go

```diff
@@ -0,0 +1,155 @@
+package deploy
+
+import (
+	"errors"
+	"fmt"
+	"net/http"
+	"strconv"
+
+	"github.com/davidebianchi/go-jsonclient"
+	"github.com/mia-platform/miactl/renderer"
+	"github.com/spf13/cobra"
+	"github.com/spf13/viper"
+)
+
+type deployConfig struct {
+	Environment         string
+	Revision            string
+	SmartDeploy         bool
+	ForceDeployNoSemVer bool
+}
+
+type deployRequest struct {
+	Environment             string `json:"environment"`
+	Revision                string `json:"revision"`
+	DeployType              string `json:"deployType"`
+	ForceDeployWhenNoSemver bool   `json:"forceDeployWhenNoSemver,omitempty"`
+}
+
+type deployResponse struct {
+	Id  int    `json:"id"`
+	Url string `json:"url"`
+}
+
+func NewDeployCmd() *cobra.Command {
+	var (
+		baseURL   string
+		apiToken  string
+		projectId string
+	)
+
+	cfg := deployConfig{}
+
+	cmd := &cobra.Command{
+		Use:   "deploy",
+		Short: "deploy project",
+		Long:  "trigger the deploy pipeline for selected project",
+		PreRunE: func(cmd *cobra.Command, args []string) error {
+			baseURL = viper.GetString("apibaseurl")
+			apiToken = viper.GetString("apitoken")
+			projectId = viper.GetString("project")
+
+			if apiToken == "" {
+				return errors.New("missing API token - please login")
+			}
+			if projectId == "" {
+				cmd.MarkFlagRequired("project")
+			}
+
+			return nil
+		},
+		RunE: func(cmd *cobra.Command, args []string) error {
+			if baseURL == "" {
+				return errors.New("API base URL not specified nor configured")
+			}
+			if projectId == "" {
+				return errors.New("project id not specified nor configured")
+			}
+			r := renderer.New(cmd.OutOrStdout())
+
+			deployData, err := deploy(baseURL, apiToken, projectId, &cfg)
+			if err != nil {
+				r.Error(err).Render()
+				return nil
+			}
+
+			// store triggered pipelines details to enabling checking their status
+			var pipelines Pipelines
+			if err := viper.UnmarshalKey(triggeredPipelinesKey, &pipelines); err != nil {
+				return err
+			}
+
+			viper.Set(triggeredPipelinesKey, append(pipelines, Pipeline{
+				ProjectId:   projectId,
+				PipelineId:  deployData.Id,
+				Environment: cfg.Environment,
+			}))
+			if err := viper.WriteConfig(); err != nil {
```

Comment:
From the requirements it was requested to use a different command (differently from console behaviour), which is the `deploy status` command (see #41). In this manner it is possible to trigger multiple deploy pipelines and then wait with a single command on all of them (only the ones triggered via `miactl` up to that moment).

Reply:
How to handle with the config file? If there are multiple runs? And when it is deleted from config?

It is possible to return the trigger id, and reuse it in input from the user?

Reply:
The pipeline id is always returned in case of successful deploy trigger, so that it would be possible to easily check its status.

However, current plan is to not offer a single pipeline check, but rather offer a command that verify the deploy status of all the pipelines previously triggered (from `miactl`). Once the status of a pipeline is checked (which mean the pipeline has completed), it is removed from the configuration.

A simple use case of this commands would be:

- _trigger deploy pipeline_
- _trigger deploy pipeline_
- _trigger deploy pipeline_
- _check deploy status_ of previous pipelines (this command should iterate over previous triggered pipelines, waiting the completion of  each one (success/failure/cancel))

In case the check _check deploy status_ command is not called, the config file would continue to accumulate triggered pipelines. The list of triggered pipelines would be emptied only when _check deploy status_ command is called (checking all the pipelines triggered till that moment).

**Note:** at the moment the command does not support concurrent executions (e.g. trigger two pipelines from two different terminals)

Reply:
Make a stateful interface in a cli is not a common use case. Furthermore, it could bring to unexpected flow.

If you think about an user which trigger some deploy without check the status and run the status check after one month, it has a useless response. 
I'd prefer to handle the status check by user, taking the trigger Id from stdout and reusing it as input of the status command (handle also a bulk request). In this way the command is still stateless, and for the use case you define:

- trigger deploy with Id 1 (Id returned by backend)
- trigger deploy with Id 2
- trigger deploy with Id 3
- miactl deploy status 1,2,3

For the use case, the bulk could be avoided at the moment. In this way you could know easily if a deploy throws error or is successful.

Reply:
I changed the implementation to reflect the use case you provided.